sled-agent/src/services.rs

```diff
@@ -623,6 +617,130 @@ impl ServiceManager {
+                    "component" => "DpdClient"
+                )),
+            },
+        );
+
+        let mut ports = vec![];
+        for svc in &req.services {
+            let external_ip;
+            let (nic, snat, external_ips) = match svc {
+                ServiceType::Nexus { external_ip, nic, .. } => {
+                    (nic, None, std::slice::from_ref(external_ip))
+                }
+                ServiceType::ExternalDns { dns_address, nic, .. } => {
+                    external_ip = dns_address.ip();
+                    (nic, None, std::slice::from_ref(&external_ip))
+                }
+                ServiceType::BoundaryNtp { nic, snat_cfg, .. } => {
+                    (nic, Some(*snat_cfg), &[][..])
+                }
+                _ => continue,
+            };
+
+            // Create the OPTE port for the service.
+            // Note we don't plumb any firewall rules at this point,
+            // Nexus will plumb them down later but the default OPTE
+            // config allows outbound access which is enough for
+            // Boundary NTP which needs to come up before Nexus.
+            let port = port_manager
+                .create_port(nic, snat, external_ips, &[])
+                .map_err(|err| Error::ServicePortCreation {
+                service: svc.to_string(),
+                err: Box::new(err),
+            })?;
+
+            // We also need to update the switch with the NAT mappings
+            let nat_target = dpd_client::types::NatTarget {
+                inner_mac: dpd_client::types::MacAddr {
+                    a: port.0.mac().into_array().to_vec(),
+                },
+                internal_ip: *underlay_address,
+                vni: port.0.vni().as_u32().into(),
+            };
+
+            let (target_ip, first_port, last_port) = match snat {
+                Some(s) => (s.ip, s.first_port, s.last_port),
+                None => (external_ips[0], 0, u16::MAX),
+            };
+
+            let nat_create = || async {
+                info!(
+                    self.inner.log, "creating NAT entry for service";
+                    "service" => ?svc,
+                );
+                match &target_ip {
+                    IpAddr::V4(ip) => {
+                        dpd_client
+                            .nat_ipv4_create(
+                                ip,
+                                first_port,
+                                last_port,
+                                &nat_target,
+                            )
+                            .await
+                    }
+                    IpAddr::V6(ip) => {
+                        dpd_client
+                            .nat_ipv6_create(
+                                ip,
+                                first_port,
+                                last_port,
+                                &nat_target,
+                            )
+                            .await
```

Comment:
We discussed this a little bit offline, but since calls to dendrite are stateful, we may want to consider placing these in a saga.

There is definitely a chicken-and-egg problem here -- on reboot:
- sled agent needs to be able to program dendrite to give NTP access to external servers...
- ... which is needed to start CRDB
- ... which is needed to start Nexus, which can run other sagas

That specific scenario aside, there may be value in making these calls from a saga, where they can be rolled back on failure. As implemented, if we fail part-way through the `for svc in &req.services` loop, we'll simply leak these resources.

Reply:
(In the interest of time, we can totally punt this to a follow-up PR, but we should track this improvement somehow)